# 빈칸이 많은 컬럼과 값이 안 변하는 컬럼 찾기

**담당: 황유진**  ·  관련 Issue: #16 (번호를 채우세요)

## 이 노트북에서 할 일

온도 CSV의 각 컬럼을 살펴보고,
**빈칸(결측)이 많은 컬럼**, **값이 계속 똑같은 컬럼**, **중복된 행**이 있는지 확인합니다.

## 왜 하는지

값이 계속 똑같은 컬럼은 분석에 아무 도움이 안 되므로 빼야 합니다.
빈칸이 많은 컬럼은 그냥 쓰면 계산이 틀어지므로, 채울지 뺄지 미리 정해야 합니다.
중복 행이 있으면 같은 시각이 두 번 세어져서 평균이나 개수가 전부 어긋납니다.

## 진행 방법

1. 아래 칸을 **위에서부터 순서대로** 실행하세요.
2. `# TODO` 라고 적힌 곳을 직접 채우세요. 막히면 디스코드에 물어보세요.
3. 맨 아래 **결과 정리** 칸에 확인한 값을 한국어 문장으로 적으세요. 이게 진짜 결과물입니다.
4. 다 했으면 커밋 전에 맨 마지막 안내를 읽으세요.

> 데이터 파일이 없으면 `data/` 폴더에 CSV 3개를 먼저 넣으세요. 이 파일들은 Git에 올라가지 않습니다(공유받은 자료를 각자 직접 넣습니다).

## 공통 준비

In [3]:
# 이 칸은 그대로 실행하세요. 데이터 경로를 잡아둡니다.
import os
import pandas as pd

DATA_DIR = os.path.join("..", "data")          # notebooks 폴더 기준 한 단계 위의 data 폴더
TEMP_CSV   = os.path.join(DATA_DIR, "T-CR1-CAL01_온도.csv")
EVENT_CSV  = os.path.join(DATA_DIR, "G-02_조업이벤트.csv")
REPAIR_CSV = os.path.join(DATA_DIR, "G-01_정기수리캘린더.csv")

pd.set_option("display.max_columns", 50)
print("경로 확인:", os.path.exists(TEMP_CSV), os.path.exists(EVENT_CSV), os.path.exists(REPAIR_CSV))

경로 확인: True True True


### 1단계 — 데이터를 읽고 전체 모양을 보세요

In [4]:
df = pd.read_csv(TEMP_CSV, encoding="utf-8")

print("행 수, 열 수:", df.shape)
df.head()

행 수, 열 수: (345390, 24)


,MEAS_DT,LOT_NO,CLN-IN,TC-Z1,TC-Z2,OP-Z1,OP-Z2,TC-Z3,TC-Z4,TC-Z5,TC-Z6,OP-Z3,OP-Z4,OP-Z5,OP-Z6,CP-Z4,CP-Z4M,TC-Z7,TC-Z8,TC-Z9,OP-Z7,OP-Z8,TC-OUT1,TC-OUT2
0,2024-01-02 03:36:00,240001,69.0111,100.559,99.7062,66.6165,18.5975,859.683,860.800,860.367,860.126,83.9075,47.2134,51.6966,69.0189,0.449234,1.151760e-10,331.202,331.335,327.550,65.4737,56.2615,291.906,287.039
1,2024-01-02 03:36:01,240001,69.0724,100.497,99.6447,73.2873,20.9099,859.621,860.737,860.429,860.126,88.8233,48.6681,50.1850,69.0560,0.449334,1.152830e-10,331.202,331.335,327.550,65.4737,56.2615,291.906,287.039
2,2024-01-02 03:36:02,240001,68.9497,100.497,99.7062,70.7377,20.9651,859.683,860.737,860.367,860.126,84.2600,48.5541,51.7643,69.0900,0.449717,1.151760e-10,331.202,331.274,327.489,65.4737,56.2615,291.845,287.039
3,2024-01-02 03:36:03,240001,68.8884,100.497,99.7062,70.4484,18.7313,859.683,860.675,860.367,860.188,84.5850,49.9122,51.7661,67.6582,0.449888,1.151760e-10,331.202,331.274,327.489,66.5701,56.2615,291.845,287.039
4,2024-01-02 03:36:04,240001,68.8884,100.436,99.7062,70.0716,21.1505,859.683,860.737,860.367,860.126,84.8836,48.2923,51.7676,70.7738,0.449992,1.151760e-10,331.202,331.214,327.550,65.4737,56.2615,291.845,287.039


### 2단계 — 컬럼별 빈칸(결측) 개수를 세세요

`isna()` 는 빈칸이면 True를 돌려줍니다. 여기에 `.sum()` 을 붙이면 개수가 나옵니다.

In [28]:
# TODO: 컬럼별 빈칸 개수를 출력하세요.
# 힌트: df.isna().sum()
empty = df.isna().sum()
print(empty)

print("===============")

# TODO: 비율(%)로도 확인해 보세요.
# 힌트: (df.isna().mean() * 100).round(2)
empty_per = (df.isna().mean() * 100).round(2)
print(empty_per)

MEAS_DT      0
LOT_NO       0
CLN-IN       4
TC-Z1       19
TC-Z2       23
OP-Z1        0
OP-Z2        0
TC-Z3        5
TC-Z4       12
TC-Z5       15
TC-Z6       16
OP-Z3      254
OP-Z4        0
OP-Z5        0
OP-Z6        0
CP-Z4        0
CP-Z4M      13
TC-Z7       22
TC-Z8       17
TC-Z9       18
OP-Z7       20
OP-Z8       17
TC-OUT1      6
TC-OUT2     13
dtype: int64
MEAS_DT    0.00
LOT_NO     0.00
CLN-IN     0.00
TC-Z1      0.01
TC-Z2      0.01
OP-Z1      0.00
OP-Z2      0.00
TC-Z3      0.00
TC-Z4      0.00
TC-Z5      0.00
TC-Z6      0.00
OP-Z3      0.07
OP-Z4      0.00
OP-Z5      0.00
OP-Z6      0.00
CP-Z4      0.00
CP-Z4M     0.00
TC-Z7      0.01
TC-Z8      0.00
TC-Z9      0.01
OP-Z7      0.01
OP-Z8      0.00
TC-OUT1    0.00
TC-OUT2    0.00
dtype: float64


### 3단계 — 값이 계속 똑같은 컬럼을 찾으세요

`nunique()` 는 그 컬럼에 **서로 다른 값이 몇 개 있는지** 세어줍니다.
이 값이 1이면 처음부터 끝까지 같은 값만 들어 있다는 뜻입니다.

In [30]:
# TODO: 컬럼별 고유값 개수를 확인하세요.
# 힌트: df.nunique()
count = df.nunique()
print(count)

print("===============")

# TODO: 고유값이 1개인 컬럼만 골라내세요.
# 힌트: const_cols = df.columns[df.nunique() <= 1]
#       print(list(const_cols))
const_cols = df.columns[df.nunique() <= 1]
print(list(const_cols))

MEAS_DT    345390
LOT_NO         14
CLN-IN      32649
TC-Z1       18693
TC-Z2       16742
OP-Z1      106210
OP-Z2      101464
TC-Z3       24883
TC-Z4       16412
TC-Z5       16050
TC-Z6       16580
OP-Z3      212963
OP-Z4      167916
OP-Z5      135260
OP-Z6      117877
CP-Z4       31882
CP-Z4M      19315
TC-Z7       13426
TC-Z8       13860
TC-Z9       23453
OP-Z7       29288
OP-Z8       25837
TC-OUT1     34632
TC-OUT2     27040
dtype: int64
[]


### 4단계 — 중복된 행이 있는지 확인하세요

두 가지를 봅니다.

1. **모든 컬럼이 똑같은 행**이 두 번 들어간 경우
2. **같은 시각(MEAS_DT)이 두 번 이상** 나온 경우 — 값이 달라도 문제입니다

In [31]:
# TODO: 완전히 똑같은 중복 행 개수를 세세요.
# 힌트: df.duplicated().sum()
same = df.duplicated().sum()
print(same)

print("===============")

# TODO: MEAS_DT 가 중복된 경우도 세세요.
# 힌트: df["MEAS_DT"].duplicated().sum()
same_time = df["MEAS_DT"].duplicated().sum()
print(same_time)

0
0


### 5단계 — 숫자 컬럼의 값 범위를 훑어보세요

온도 컬럼(`TC-Z1` ~ `TC-Z6` 등)에 말이 안 되는 값(예: 음수, 0, 아주 큰 값)이 있는지 봅니다.

In [32]:
# TODO: 숫자 컬럼들의 요약 통계를 확인하세요.
# 힌트: df.describe().T
df[["TC-Z1", "TC-Z2", "TC-Z3", "TC-Z4", "TC-Z5", "TC-Z6"]].describe().T
# 눈에 띄는 값이 있으면 아래 결과 정리에 적어주세요.

,count,mean,std,min,25%,50%,75%,max
TC-Z1,345371.0,99.923794,0.568980,95.5681,99.80200,99.9889,100.163,102.076
TC-Z2,345367.0,99.938705,0.491607,95.4042,99.88195,99.9980,100.108,101.576
TC-Z3,345385.0,859.237644,5.333726,812.4740,859.02700,860.2460,861.028,872.549
TC-Z4,345378.0,859.266329,4.660903,814.0570,859.73500,860.1360,860.425,862.255
TC-Z5,345375.0,859.283996,4.642771,814.2280,859.75500,860.1860,860.435,861.556
TC-Z6,345374.0,859.298111,4.662084,813.9670,859.75400,860.1630,860.462,862.507


## 결과 정리 — 여기를 꼭 채우세요

아래 빈칸을 확인한 값으로 바꿔서 적으세요. 이 내용이 `docs/02-data-contract.md`로 옮겨집니다.

| 항목 | 확인한 값 |
| --- | --- |
| 전체 행 수 / 컬럼 수 | 345390 행 / 24 열 |
| 빈칸이 있는 컬럼 | CLN-IN, TC-Z1, TC-Z2, TC-Z3, TC-Z4, TC-Z5, TC-Z6, OP-Z3, CP-Z4M, TC-Z7, TC-Z8, TC-Z9, OP-Z7, OP-Z8, TC-OUT1, TC-OUT2 (없으면 "없음") |
| 빈칸 비율이 가장 높은 컬럼 | OP-Z3 (0.07 %) |
| 값이 계속 똑같은 컬럼 | 없음 |
| 완전히 똑같은 중복 행 | 0 개 |
| 같은 시각(MEAS_DT)이 두 번 이상 나온 경우 | 0 개 |


### 이상하다고 느낀 점 / 확실하지 않은 점

- (있으면 적어주세요. 없으면 "없음"이라고 적으세요.)
- 없음
---

## 커밋하기 전에 읽으세요

1. **출력은 지우지 않아도 됩니다.** `nbstripout` 필터를 등록해 뒀다면 `git add` 할 때 자동으로 지워집니다.
   등록했는지 확인: `python -m nbstripout --status` → `Automatic cleanup enabled` 가 나와야 합니다.
   안 나오면: `python -m nbstripout --install --attributes .gitattributes`
2. 이 노트북 **한 파일만** 커밋하세요. 다른 사람 파일은 건드리지 마세요.
3. 브랜치를 만들어서 작업하세요. 예) `git switch -c feat/이슈번호-설명`
4. 커밋 메시지 첫 줄: `feat: 결측·상수 컬럼과 중복 행 확인`
5. PR 본문에 `Closes #이슈번호` 를 꼭 적으세요.